### PR TITLE
deduplicate `splint` in FITPACK wrappers; take 3

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

A third alternative to gh-16871 and gh-16932. See https://github.com/scipy/scipy/issues/16729 for context.

#### What does this implement/fix?

This continues @AtsushiSakai 's removal of `splint` from the handwritten `_fitpack` extentsion : this is one of the routines which has two sets of wrappers: handwritten in `_fitpack` and f2py-generated in `dfitpack`. Over at gh-16932 @rgommers suggested to push array size checking from python to f2py layer, which this PR does at an expense of a minor Fortran surgery.

#### Additional information

I do not insist on that this approach being the right one; only one of gh-16971, gh-16932 and this one would go in (or none if we decide to just leave it be).

WDYT @AtsushiSakai @rgommers, others?

### CODE REVIEW

File: scipy/interpolate/fitpack/splint.f [hunk 1]
```diff
@@ -41,9 +43,9 @@ recursive function splint(t,n,c,k,a,b,wrk) result(splint_res)
 c
 c  ..scalar arguments..
       real*8 a,b
-      integer n,k
+      integer n,k, nc
 c  ..array arguments..
-      real*8 t(n),c(n),wrk(n)
+      real*8 t(n),c(nc),wrk(n)
```
Comment (ev-br) [inline scipy/interpolate/fitpack/splint.f:48]: To verify that this, compare to the  `do i=1, nk1` loop below, where `nk1=n-k-1`. This way, the elemnts of the `c` array with indices past `n-k-1` are not referenced even though the array is declated to be of size `n`.

**Review by AtsushiSakai — APPROVED** [review]
LGTM. I think this option is better than https://github.com/scipy/scipy/pull/16871 and https://github.com/scipy/scipy/pull/16932, because python codes look simpler.

**Review by rgommers — APPROVED** [review]
This looks like the cleanest PR of the three alternatives so far, so happy for this one to move ahead.

Note that I didn't verify the f2py signature file change, that always takes me quite a while as a non-expert. If you're confident it's right, I'd say go ahead and merge this @ev-br.

**Comment by ev-br** [comment]
Not sure if *confident* and f2py  fit well into the same sentence :-). That said, I do believe the current state of this patch is right to the best of my knowledge and understanding. If there's a problem, I don't see anything that can't be fixed or reverted. 

Thanks for taking a look @AtsushiSakai and @rgommers , and special thank you to @AtsushiSakai for the bulk of the work in these three PRs!

**Pull request merged**